RecoLocalMuon/GEMRecHit/src/ME0SegAlgoMM.cc

```diff
@@ -0,0 +1,551 @@
+    float phi_new = theEnsemble.first->toGlobal(newChain[iRH_new]->localPosition()).phi();
+    float eta_new = theEnsemble.first->toGlobal(newChain[iRH_new]->localPosition()).eta();
+    for(size_t iRH_old = 0;iRH_old<oldChain.size();++iRH_old){      
+      int layer_old = oldChain[iRH_old]->me0Id().layer();
+      float phi_old = theEnsemble.first->toGlobal(oldChain[iRH_old]->localPosition()).phi();
+      float eta_old = theEnsemble.first->toGlobal(oldChain[iRH_old]->localPosition()).eta();
+      // to be chained, two hits need to be in neighbouring layers...
+      // or better allow few missing layers (upto 3 to avoid inefficiencies);
+      // however we'll not make an angle correction because it
+      // worsen the situation in some of the "regular" cases 
+      // (not making the correction means that the conditions for
+      // forming a cluster are different if we have missing layers -
+      // this could affect events at the boundaries ) 
+      //to be chained, two hits need also to be "close" in phi and eta
+      bool layerRequirementOK = abs(layer_new-layer_old)<5;
+      bool phiRequirementOK = fabs(phi_old-phi_new) < dPhiChainBoxMax;
+      bool etaRequirementOK = fabs(eta_old-eta_new) < dEtaChainBoxMax;
+      
+      if(layerRequirementOK && phiRequirementOK && etaRequirementOK){
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+
+
+
+
+std::vector<ME0Segment> ME0SegAlgoMM::buildSegments(const EnsembleHitContainer& rechits) {
+  std::vector<ME0Segment> me0segs;
+
+  proto_segment.clear();
+  // select hits from the ensemble and sort it 
+  for (auto rh=rechits.begin(); rh!=rechits.end();rh++){
+    proto_segment.push_back(*rh);
+  }
+  if (proto_segment.size() < minHitsPerSegment){
+    return me0segs;
+  }
+  // The actual fit on all hit of the protosegments;
+  this->doSlopesAndChi2();
+  this->fillLocalDirection();
+  AlgebraicSymMatrix protoErrors = this->calculateError();
+  this->flipErrors( protoErrors ); 
+  ME0Segment tmp(proto_segment,protoIntercept, protoDirection, protoErrors,protoChi2);
+  me0segs.push_back(tmp);
+  return me0segs;
+}
+
+//Method doSlopesAndChi2
+// fitSlopes() and  fillChiSquared() are always called one after the other 
+// In fact the code is duplicated in the two functions (as we need 2 loops) - 
+// it is much better to fix that at some point 
+void ME0SegAlgoMM::doSlopesAndChi2(){
+  this->fitSlopes();
+  this->fillChiSquared();
+}
+/* Method fitSlopes
+ *
+ * Perform a Least Square Fit on a segment as per SK algo
+ *
+ */
+void ME0SegAlgoMM::fitSlopes() {
+
+  CLHEP::HepMatrix M(4,4,0);
+  CLHEP::HepVector B(4,0);
+  // In absence of a geometrical construction of the ME0Ensemble take layer 1  
+  const ME0EtaPartition* ens = theEnsemble.first;
+
+  for (auto ih = proto_segment.begin(); ih != proto_segment.end(); ++ih) {
+    const ME0RecHit& hit = (**ih);
+    const ME0EtaPartition* roll  = theEnsemble.second[hit.me0Id()];
```

Comment:
Still pending: Please check that "roll" is non-null before dereferencing it. This comment applies to the methods derivativeMatrix and fillChiSquared, too.